lib/TsconfigPathsPlugin.js

```diff
@@ -0,0 +1,238 @@
+
+/**
+ * @typedef {object} TsconfigPathsPluginOptions
+ * @property {string=} configFile - Path to tsconfig.json file
+ * @property {string=} baseUrl - Override baseUrl from tsconfig.json
+ * @property {string=} context - Context directory for resolving relative paths
+ */
+
+module.exports = class TsconfigPathsPlugin {
+	/**
+	 * @param {string} configFile tsconfig file path
+	 */
+	constructor(configFile) {
+		this.configFile = configFile;
+	}
+
+	/**
+	 * @param {Resolver} resolver the resolver
+	 * @returns {void}
+	 */
+	apply(resolver) {
+		if (!this.configFile) return;
+
+		const aliasTarget = resolver.ensureHook("internal-resolve");
+		const moduleTarget = resolver.ensureHook("module");
+		const aliasOptionsPromise = this.loadAndConvertPaths(resolver);
+
+		resolver
+			.getHook("raw-resolve")
+			.tapAsync(
+				"TsconfigPathsPlugin",
+				async (request, resolveContext, callback) => {
+					try {
+						const aliasOptions = await aliasOptionsPromise;
+						aliasResolveHandler(
+							resolver,
+							aliasOptions.filter((option) => option.name !== "*"),
+							aliasTarget,
+							request,
+							resolveContext,
+							callback,
+						);
+					} catch (err) {
+						callback(/** @type {Error} */ (err));
+					}
+				},
+			);
+
+		resolver
+			.getHook("raw-module")
+			.tapAsync(
+				"TsconfigPathsPlugin",
+				async (request, resolveContext, callback) => {
+					try {
+						const aliasOptions = await aliasOptionsPromise;
+						const directories = aliasOptions
+							.filter((option) => option.name === "*")
+							.map((option) => /** @type {string} */ (option.alias));
+						if (directories.length === 0) return callback();
+
+						modulesResolveHandler(
+							resolver,
+							directories,
+							moduleTarget,
+							request,
+							resolveContext,
+							callback,
+						);
+					} catch (err) {
+						callback(/** @type {Error} */ (err));
+					}
+				},
+			);
+	}
+
+	/**
+	 * Load tsconfig.json (and referenced tsconfigs) and convert paths to AliasPlugin format
+	 * @param {Resolver} resolver the resolver
+	 * @returns {Promise<AliasOption[]>} Array of alias options for AliasPlugin
+	 */
+	async loadAndConvertPaths(resolver) {
+		try {
+			const fs = resolver.fileSystem;
+			const configPath = path.isAbsolute(this.configFile)
+				? this.configFile
+				: resolver.join(process.cwd(), this.configFile);
+
+			const mainOptions = await this.readTsconfigCompilerOptions(
+				resolver,
+				configPath,
+			);
+			if (!mainOptions) return [];
+
+			/** @type {AliasOption[]} */
+			const aliases = [];
+			aliases.push(
+				...this.convertPathsToAliases(
+					resolver,
+					mainOptions.paths,
+					mainOptions.baseUrl,
+				),
+			);
+
+			// Collect references from the main tsconfig.json
+			const tsconfigJson = await new Promise((resolve, reject) => {
+				fs.readFile(configPath, "utf8", (err, data) => {
+					if (err) reject(err);
+					resolve(JSON.parse(/** @type {string} */ (data)));
+				});
+			});
```

Comment:
We are reading configuration file on the each resolve, it can have a bad performance